util/test/demos/vk/vk_blend.cpp

```diff
@@ -0,0 +1,250 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "vk_test.h"
+
+RD_TEST(VK_Blend, VulkanGraphicsTest)
+{
+  static constexpr const char *Description =
+      "Draws a triangle repeatedly to test blending within a single drawcall";
+
+  const DefaultA2V TemplateTriangleRed[3] = {
+      {Vec3f(-0.5f, -0.5f, 0.0f), Vec4f(1 / 255.f, 0.0f, 0.0f, 1.0f), Vec2f(0.0f, 0.0f)},
+      {Vec3f(0.0f, 0.5f, 0.0f), Vec4f(1 / 255.f, 0.0f, 0.0f, 1.0f), Vec2f(0.0f, 1.0f)},
+      {Vec3f(0.5f, -0.5f, 0.0f), Vec4f(1 / 255.f, 0.0f, 0.0f, 1.0f), Vec2f(1.0f, 0.0f)},
+  };
+  const int TRIANGLES_RED_INDEX = 0;
+  const int NUM_TRIANGLES_RED = 16;
+  const DefaultA2V TemplateTriangleGreen[3] = {
+      {Vec3f(-0.5f, -0.5f, 0.0f), Vec4f(0.0f, 1 / 255.f, 0.0f, 1.0f), Vec2f(0.0f, 0.0f)},
+      {Vec3f(0.0f, 0.5f, 0.0f), Vec4f(0.0f, 1 / 255.f, 0.0f, 1.0f), Vec2f(0.0f, 1.0f)},
+      {Vec3f(0.5f, -0.5f, 0.0f), Vec4f(0.0f, 1 / 255.f, 0.0f, 1.0f), Vec2f(1.0f, 0.0f)},
+  };
+  const int TRIANGLES_GREEN_INDEX = TRIANGLES_RED_INDEX + NUM_TRIANGLES_RED;
+  const int NUM_TRIANGLES_GREEN = 255;
+  const DefaultA2V TemplateTriangleBlue[3] = {
+      {Vec3f(-0.5f, -0.5f, 0.0f), Vec4f(0.0f, 0.0f, 1 / 255.f, 1.0f), Vec2f(0.0f, 0.0f)},
+      {Vec3f(0.0f, 0.5f, 0.0f), Vec4f(0.0f, 0.0f, 1 / 255.f, 1.0f), Vec2f(0.0f, 1.0f)},
+      {Vec3f(0.5f, -0.5f, 0.0f), Vec4f(0.0f, 0.0f, 1 / 255.f, 1.0f), Vec2f(1.0f, 0.0f)},
+  };
+  const int TRIANGLES_BLUE_INDEX = TRIANGLES_GREEN_INDEX + NUM_TRIANGLES_GREEN;
+  const int NUM_TRIANGLES_BLUE = 512;
+
+  const int NUM_TRIANGLES_TOTAL = TRIANGLES_BLUE_INDEX + NUM_TRIANGLES_BLUE;
+
+  int main()
+  {
+    // Blending on an SRGB color space gives incorrect results for this test,
+    // as eventually the values it outputs are too small to make a difference.
+    // Use a linear color space instead.
+    requestedSwapChainFormat = VK_FORMAT_B8G8R8A8_UNORM;
```

Comment:
I've changed it to RGBA32F but left the values at 1.0/255.0 for now, and changed the test to handle that (which only requires treating the last value specially). But, currently the presented image shows up as black/undefined, which doesn't affect the test itself but still might be confusing. (`1.0/255.0` seems better for showing it on the screen directly, though I'm sure it's also possible to normalize it afterwards; I just didn't find a helper for that.)

I'm not sure how `VK_Shader_Debug_Zoo` handles that (though also, I couldn't run it on my machine as the demo isn't usable due to "Limit 'maxPerStageDescriptorStorageImages' 16 is insufficient (need at least 64)").

Reply:
Yeh `VK_Shader_Debug_Zoo` doesn't render to the backbuffer, it just clears it. Since it's only rendering single pixels there's not much to output there and all the actual python checks are on the offscreen target. Some tests like `VK_CBuffer_Zoo` will use the `blitToSwap()` helper to blit to the backbuffer, which you can use if you want to have a preview.

16 for maxPerStageDescriptorStorageImages sounds absolutely _crippled_. Are you running on a software emulation vulkan ICD, or a many-years-old vulkan 1.0 minimum spec device?

Reply:
Thanks, `blitToSwap` did what I wanted.

> 16 for maxPerStageDescriptorStorageImages sounds absolutely crippled. Are you running on a software emulation vulkan ICD, or a many-years-old vulkan 1.0 minimum spec device?

It's an Intel integrated GPU. It does support Vulkan 1.3, but it's generally fairly weak.